### ISSUE TITLE
HTMLElement.prototype.offsetParent leaks a node inside a shadow tree

### ISSUE THREAD

**Issue by rniwa** [WICG/webcomponents#497]
We need to patch `offsetParent` so that when a node's container block is inside a shadow tree, we don't leak a node inside a shadow tree.

Chrome seems to return `null` in such cases.  It's somewhat undesirable in that this makes the feature useless (since you can't determine where your containing block is) but I don't have a strong opinion about this as long as we don't leak a node inside a shadow tree.

**Comment by rniwa** [comment]
@annevk @zcorpan @hayatoito @travisleithead

**Comment by rniwa** [comment]
It looks like Chrome returns `null` whenever the offset parent is inside a shadow tree.  I think a better change would be to return `null` if the offset parent is not an [unclosed node](http://w3c.github.io/webcomponents/spec/shadow/#dfn-unclosed-node) of the context object.

**Comment by rniwa** [comment]
An alternative solution is to walk up `offsetParent` enough to find the first offset parent that is [unclosed node](http://w3c.github.io/webcomponents/spec/shadow/#dfn-unclosed-node) to the context object.

This is probably a better fix since authors won't be able to determine the location of an element which is assigned to a slot otherwise.

**Comment by kojiishi** [comment]
Should `offsetLeft` and `offsetTop` be adjusted accordingly?

**Comment by rniwa** [comment]
Right. Since [offsetTop](https://drafts.csswg.org/cssom-view/#dom-htmlelement-offsettop) and [offsetLeft](https://drafts.csswg.org/cssom-view/#dom-htmlelement-offsetleft) are defined to use [offsetParent](https://drafts.csswg.org/cssom-view/#dom-htmlelement-offsetparent) already, that's kind of automatic.

**Comment by hayatoito** [comment]
I have filed an issue for csswg: https://github.com/w3c/csswg-drafts/issues/159

**Comment by hayatoito** [comment]
PR is merged. Let me close this issue.

**Comment by rniwa** [comment]
Sorry, it looks like I missed an crucial point in https://github.com/w3c/csswg-drafts/issues/159

We’re exposing any node that’s not closed-shadow-hidden here, which means we’d return an element inside shadow tree as an offset parent. I don’t think we want that even in an open shadow tree.

Let’s imagine you’re trying to traverse over ancestor offsetParents. Such a code would now walk over elements inside a shadow tree. I don’t think this is desirable just as much as finding nodes via querySelector or getElementsByTagName inside shadow trees are not desirable.

**Comment by rniwa** [comment]
Rough consensus at TPAC F2F: Use the [retargeting algorithm](https://dom.spec.whatwg.org/#retarget) to find the offset parent, and the compute left and top from that node.

**Comment by rniwa** [comment]
We need to resolve CSS OM View spec change in https://github.com/w3c/csswg-drafts/issues/159

**Comment by eKoopmans** [comment]
**Note:** This behaviour appears to have now been updated in Chrome 109 (and earlier versions of Firefox/Webkit) so that it no longer leaks elements within a shadow DOM.

If you need the original behaviour, see [this comment from October 2022](https://bugs.chromium.org/p/chromium/issues/detail?id=1331803#c9) and the linked polyfill:
https://github.com/josepharhar/offsetparent-polyfills
